File: Node/Euler011.js
#!/usr/bin/env node
const fs = require("fs")

function main() {
  const target = 4;
  table = loadFile("input/Euler011.txt");
  biggest = maxProduct(table, target);
  console.log(biggest);
}

function loadFile(path) {
  var contents = fs.readFileSync(path, 'utf8');
  table = contents.split("\n").filter((line) => line).map((line) => {
    return line.split(" ").filter((item) => item).map((item) => parseInt(item));
  });
  return table;
}

function maxProduct(table, length) {
  biggest = 1;

  for (let i = 0; i < table.length; i++) {
    for (let j = length - 1; j < table[0].length; j++) {
      var horizProduct = 1;
      var vertProduct = 1;
      var diagEastProduct= 1;
      var diagWestProduct = 1;

      for (let cursor = 0; cursor < length; cursor++) {
        horizProduct *= table[i][j-cursor];
        vertProduct *= table[j-cursor][i];
        if (i >= length) {
          diagWestProduct *= table[i-cursor][j-cursor];
          if (j <= table[0].length - length) {
            diagEastProduct *= table[i-cursor][j+cursor];
          }
        }
      }

      biggest = Math.max(biggest, horizProduct, vertProduct, diagEastProduct, diagWestProduct);
    }
  }

  return biggest;
}

if (require.main === module) {
  main();
}

File: Kotlin/Euler011.kt
package Euler011

import java.io.File

fun main() {
  val target = 4
  val table = loadFile("input/Euler011.txt")
  println(maxProduct(table, target))
}

fun loadFile(path: String): List<List<Int>> {
  val table: MutableList<List<Int>> = mutableListOf()

  File(path).forEachLine { line ->
    table.add(line.split(" ").asSequence().filter { numStr -> numStr != "" }.map { numStr -> numStr.toInt() }.toList())
  }

  return table
}

fun maxProduct(table: List<List<Int>>, length: Int): Int {
  var biggest = -1
  for (i in 0..(table.size - 1)) {
    for (j in (length - 1)..(table[i].size - 1)) {
      val horizProduct = table[i].slice((j-length+1)..j).reduce { a, b -> a * b }
      val vertProduct = table.slice((j-length+1)..j).map { row -> row[i] }.reduce { a, b -> a * b }

      var diagWest = -1
      var diagEast = -1
      if (i >= length) {
        diagWest = (0..(length-1)).map { cursor -> table[i - cursor][j - cursor] }.reduce { a, b -> a * b }
        if (table[i].size > j + 4) {
          diagEast = (0..(length-1)).map { cursor -> table[i - cursor][j + cursor] }.reduce { a, b -> a * b }
        }
      }

      biggest = sequenceOf(biggest, horizProduct, vertProduct, diagEast, diagWest).max() as Int
    }
  }

  return biggest
}
